File: solutions/Sliding Window/3_longest-substring-without-repeating-characters.py
from collections import deque

class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        mp = {}
        l, r = 0, 0
        res = 0

        for r in range(len(s)):
            if s[r] in mp:
                l = max(mp[s[r]] + 1, l)

            res = max(res, r - l + 1)
            mp[s[r]] = r
 
        return res

# Time Complexity: O(n)
# Space Complexity: O(n)


class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        l, r = 0, 0
        res = 0
        seen = set()

        for r in range(len(s)):
            while s[r] in seen:
                seen.remove(s[l])
                l += 1

            res = max(res, r - l + 1)
            seen.add(s[r])

        return res

# Time Complexity: O(n)
# Space Complexity: O(n)


class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        l, r = 0, 0
        res = 0
        seen = deque()

        for r in range(len(s)):
            while s[r] in seen:
                seen.popleft()
                l += 1

            res = max(res, r - l + 1)
            seen.append(s[r])

        return res

# Time Complexity: O(n)
# Space Complexity: O(n)

File: C++/Sliding Window/3_longest-substring-without-repeating-characters.cpp
#include <unordered_set>
#include <unordered_map>
#include <string>

using namespace std;

class Solution {
public:
    int lengthOfLongestSubstring(string s) {
        size_t n = s.size();
        unordered_set<char> seen;
        int res = 0;

        int l = 0;
        for (int r = 0; r < n; r++) {
            while (seen.contains(s[r])) {
                seen.erase(s[l]);
                l++;
            }
            
            seen.insert(s[r]);
            res = max(res, r - l + 1);
        }

        return res;
    }
};


class Solution {
public:
    int lengthOfLongestSubstring(string s) {
        size_t n = s.size();
        unordered_map<char, int> map;
        int res = 0;

        int l = 0;
        for (int r = 0; r < n; r++) {
            if (map.contains(s[r])) {
                l = max(l, map[s[r]] + 1);
            }
            map[s[r]] = r;
            res = max(res, r - l + 1);
        }

        return res;
    }
};